src/relay/pass/fold_conv2d.cc

```diff
@@ -0,0 +1,163 @@
+};
+
+std::tuple<Expr, IndexExpr> TransformWeight(std::vector<const CallNode*> convolutions) {
+  int64_t num_filters = 0;  // number of filters of the transformed weight
+  Array<Expr> weights;
+  for (const CallNode* n : convolutions) {
+    weights.push_back(n->args[1]);
+    auto channels = as_const_int(n->attrs.as<Conv2DAttrs>()->channels);
+    CHECK(channels);
+    num_filters += *channels;
+  }
+  return std::tuple<Expr, IndexExpr>{MakeConcatenate(TupleNode::make(weights), 0),
+                                     MakeConstScalar(Int(32), num_filters)};
+}
+
+// Two 2d convolutions can be combined if they have the same attributes or only have
+// different output channels.
+bool IsCompatibleConv2D(const Conv2DAttrs& a, const Conv2DAttrs& b) {
+  AttrsEqual eq;
+  return eq(a.strides, b.strides) &&
+         eq(a.padding, b.padding) &&
+         eq(a.dilation, b.dilation) &&
+         eq(a.groups, b.groups) &&
+         eq(a.kernel_size, b.kernel_size) &&
+         eq(a.data_layout, b.data_layout) &&
+         eq(a.weight_layout, b.weight_layout) &&
+         eq(a.out_dtype, b.out_dtype) &&
+         eq(a.out_layout, b.out_layout);
+}
+
+Expr MakeFoldedConv2D(const Expr& data, const std::vector<const CallNode*>& convolutions) {
+  static const Op& conv2d = Op::Get("nn.conv2d");
+
+  Expr new_weight;
+  IndexExpr new_channels;
+  std::tie(new_weight, new_channels) = TransformWeight(convolutions);
+
+  const CallNode* group_root = *(convolutions).begin();
+  auto attrs = group_root->attrs.as<Conv2DAttrs>();
+  auto new_attrs = make_node<Conv2DAttrs>();
+  new_attrs->strides = attrs->strides;
+  new_attrs->padding = attrs->padding;
+  new_attrs->dilation = attrs->dilation;
+  new_attrs->groups = attrs->groups;
+  new_attrs->kernel_size = attrs->kernel_size;
+  new_attrs->data_layout = attrs->data_layout;
+  new_attrs->weight_layout = attrs->weight_layout;
+  new_attrs->out_layout = attrs->out_layout;
+  new_attrs->out_dtype = attrs->out_dtype;
+  new_attrs->channels = new_channels;
+
+  return CallNode::make(conv2d, {data, new_weight}, Attrs{new_attrs}, {});
+}
+
+Expr FoldConv2D(const Expr& expr) {
+  // data -> array of conv2d with the same input
+  auto children_map = SiblingConv2DFinder().Find(expr);
+  Map<Expr, Expr> subst_map;
+
+  for (const auto& pair : children_map) {
+    Expr data = pair.first;
+    std::vector<const CallNode*> children = pair.second;
+
+    if (children.size() < 2) continue;
+
+    std::vector<size_t> group_ids(children.size());
+    std::vector<std::vector<const CallNode*>> groups;
+
+    for (size_t i = 0; i < children.size(); i++) {
+      const CallNode* n = children[i];
+      auto args = n->attrs.as<Conv2DAttrs>();
+
+      // assign a group id or create a new group for each conv2d
+      auto it = std::find_if(groups.begin(), groups.end(),
+                             [&](const std::vector<const CallNode*>& group) {
```

Comment:
can ```group``` be empty?

Reply:
@masahi `group` should have at least one element
see https://github.com/dmlc/tvm/pull/2089/files#diff-356795e1ec99fed57994eedc6806ba2eR130